ui/components/app/snaps/snap-permissions-list/snap-permissions-list.js

```diff
@@ -1,52 +1,129 @@
-import React from 'react';
+import React, { useState, useEffect, useMemo } from 'react';
 import PropTypes from 'prop-types';
 import { useSelector } from 'react-redux';
-import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 import { useI18nContext } from '../../../../hooks/useI18nContext';
-import { Box } from '../../../component-library';
+import { Box, ButtonLink } from '../../../component-library';
 import {
   getMultipleTargetsSubjectMetadata,
   getSnapsMetadata,
 } from '../../../../selectors';
+import {
+  Display,
+  FlexDirection,
+  JustifyContent,
+} from '../../../../helpers/constants/design-system';
+import SnapPermissionAdapter from '../snap-permission-adapter';
+import { PermissionWeightThreshold } from '../../../../../shared/constants/permissions';
 import { getSnapName } from '../../../../helpers/utils/util';
-import SnapPermissionCell from '../snap-permission-cell';
+import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 
 export default function SnapPermissionsList({
   snapId,
   snapName,
   permissions,
   connections,
   showOptions,
+  turnOffAbstraction,
+  showAllPermissions,
 }) {
   const t = useI18nContext();
-  const snapsMetadata = useSelector(getSnapsMetadata);
-  const permissionsToShow = {
-    ...permissions,
-    connection_permission: connections ?? {},
-  };
+
+  const combinedPermissions = useMemo(() => {
+    return { ...permissions, connection_permission: connections ?? {} };
+  }, [permissions, connections]);
+
   const targetSubjectsMetadata = useSelector((state) =>
     getMultipleTargetsSubjectMetadata(state, connections),
   );
 
-  return (
-    <Box className="snap-permissions-list">
-      {getWeightedPermissions({
+  const snapsMetadata = useSelector(getSnapsMetadata);
+
+  const [showAll, setShowAll] = useState(false);
+
+  const [permissionsToDisplay, setPermissionsToDisplay] = useState([]);
+
+  const [weightedPermissions, setWeightedPermissions] = useState({});
+
+  useEffect(() => {
```

Comment:
Why the need for `useEffect` here? It feels redundant, can this not be done without the hook? I think the only state you would need here is `showAll`.

Reply:
I was thinking of just using `turnOffAbstraction` and the `showAll` state to figure out a condition to show the correct component outside of the `useEffect`. I'm not completely understanding the need for the hook, if the permissions are memoized at the top level then what re-rendering would happen? I think you're getting extra renders by setting state within the `useEffect` as well.

Reply:
The problem is that now UI is interactive. Permission set can be changed by user action. Clicking "Show all permissions" require changing the set of permissions displayed, then re-rendering is required. So, the permission data displayed is not static, it becomes dynamic because of the user action. Here comes the need for `useEffect`. 

Without using `useEffect` UI cannot render correct permission set (from what I've tried so far). The rest of the functions related to the permission set, flags and states are added inside `useEffect` because of dependencies and re-rendering needed.